storm-core/src/jvm/org/apache/storm/daemon/supervisor/SyncSupervisorEvent.java

```diff
@@ -0,0 +1,626 @@
+            if (Utils.getBoolean(conf.get(Config.SUPERVISOR_RUN_WORKER_AS_USER), false)) {
+                SupervisorUtils.rmrAsUser(conf, stormId, path);
+            } else {
+                Utils.forceDelete(ConfigUtils.supervisorStormDistRoot(conf, stormId));
+            }
+        } catch (Exception e) {
+            LOG.info("Exception removing: {} ", stormId, e);
+        }
+    }
+
+    /**
+     * Check for the files exists to avoid supervisor crashing Also makes sure there is no necessity for locking"
+     * 
+     * @param conf
+     * @param localizer
+     * @param assignedStormIds
+     * @param allDownloadedTopologyIds
+     * @return
+     */
+    protected Set<String> verifyDownloadedFiles(Map conf, Localizer localizer, Set<String> assignedStormIds, Set<String> allDownloadedTopologyIds)
+            throws IOException {
+        Set<String> srashStormIds = new HashSet<>();
+        for (String stormId : allDownloadedTopologyIds) {
+            if (assignedStormIds.contains(stormId)) {
+                if (!SupervisorUtils.doRequiredTopoFilesExist(conf, stormId)) {
+                    LOG.debug("Files not present in topology directory");
+                    rmTopoFiles(conf, stormId, localizer, false);
+                    srashStormIds.add(stormId);
+                }
+            }
+        }
+        return srashStormIds;
+    }
+
+    /**
+     * download code ; two cluster mode: local and distributed
+     *
+     * @param conf
+     * @param stormId
+     * @param masterCodeDir
+     * @throws IOException
+     */
+    private void downloadStormCode(Map conf, String stormId, String masterCodeDir, Localizer localizer) throws Exception {
+        String clusterMode = ConfigUtils.clusterMode(conf);
+
+        if (clusterMode.endsWith("distributed")) {
+            downloadDistributeStormCode(conf, stormId, masterCodeDir, localizer);
+        } else if (clusterMode.endsWith("local")) {
+            downloadLocalStormCode(conf, stormId, masterCodeDir, localizer);
+        }
+    }
+
+    private void downloadLocalStormCode(Map conf, String stormId, String masterCodeDir, Localizer localizer) throws Exception {
+
+        String tmproot = ConfigUtils.supervisorTmpDir(conf) + Utils.FILE_PATH_SEPARATOR + Utils.uuid();
+        String stormroot = ConfigUtils.supervisorStormDistRoot(conf, stormId);
+        BlobStore blobStore = Utils.getNimbusBlobStore(conf, masterCodeDir, null);
+        try {
+            FileUtils.forceMkdir(new File(tmproot));
+            String stormCodeKey = ConfigUtils.masterStormCodeKey(stormId);
+            String stormConfKey = ConfigUtils.masterStormConfKey(stormId);
+            String codePath = ConfigUtils.supervisorStormCodePath(tmproot);
+            String confPath = ConfigUtils.supervisorStormConfPath(tmproot);
+            blobStore.readBlobTo(stormCodeKey, new FileOutputStream(codePath), null);
+            blobStore.readBlobTo(stormConfKey, new FileOutputStream(confPath), null);
```

Comment:
The `FileOutputStream` instances here are not being closed.  This can cause a lot of problems on Windows.